service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/XpathExtractor.java

```diff
@@ -0,0 +1,344 @@
+            .participant2(DOCUMENT_REFERENCE_SECOND_PARTICIPANT)
+            .build();
+
+        return extractValues(component, ehrTemplateArgs);
+    }
+
+    @SneakyThrows
+    private static EncounterTemplateParametersBuilder extractValues(String component, EhrTemplateArgs ehrTemplateArgs) {
+        Document xmlDocument = X_PATH_SERVICE.parseDocumentFromXml(component);
+        var builder = EncounterTemplateParameters.builder();
+
+        getNodeValueOptional(xmlDocument, ehrTemplateArgs.getAvailabilityTime())
+            .map(XpathExtractor::buildAvailabilityTimeTemplate)
+            .ifPresentOrElse(builder::availabilityTime,
+                () -> builder.availabilityTime(DEFAULT_AVAILABILITY_TIME_VALUE));
+
+        getNodeValueOptional(xmlDocument, ehrTemplateArgs.getAuthorTime())
+            .ifPresentOrElse(builder::authorTime,
+                () -> builder.availabilityTime(DEFAULT_AVAILABILITY_TIME_VALUE));
+
+        getNodeValueOptional(xmlDocument, ehrTemplateArgs.getEffectiveTime(), ehrTemplateArgs.getEffectiveTimeBackup())
+            .map(XpathExtractor::buildEffectiveTimeTemplate)
+            .ifPresentOrElse(builder::effectiveTime,
+                () -> builder.effectiveTime(DEFAULT_TIME_VALUE));
+
+        getNodeValueOptional(xmlDocument, ehrTemplateArgs.getAuthorAgentRef())
+            .ifPresent(builder::agentRef);
+
+        getNodeValueOptional(xmlDocument, ehrTemplateArgs.getParticipant2())
+            .map(XpathExtractor::buildParticipant)
+            .ifPresent(builder::particpant2);
+
+        return builder;
+    }
+
+    private static Optional<String> getNodeValueOptional(Document document, String firstExpression, String secondExpression) {
+        var extracted = X_PATH_SERVICE.getNodeValue(document, firstExpression, secondExpression);
+        return buildOptional(extracted);
+    }
+
+    private static Optional<String> getNodeValueOptional(Document document, String expression) {
+        var extracted = X_PATH_SERVICE.getNodeValue(document, expression);
+        return buildOptional(extracted);
+    }
+
+    private static Optional<String> buildOptional(String inputString) {
+        if (StringUtils.isNotBlank(inputString)) {
+            return Optional.of(inputString);
+        }
+        return Optional.empty();
+    }
+
+    private static String buildAvailabilityTimeTemplate(String time) {
+        return String.format(AVAILABILITY_TIME_VALUE_TEMPLATE, time);
+    }
+
+    private static String buildEffectiveTimeTemplate(String time) {
+        return String.format(EFFECTIVE_TIME_CENTER_TEMPLATE, time);
+    }
+
+    private static String buildParticipant(String participantId) {
+        return String.format(PARTICIPANT_TEMPLATE, participantId);
+    }
+
+    @Data
+    @Builder
+    private static class EhrTemplateArgs {
+        private String availabilityTime;
+        private String effectiveTime;
+        private String effectiveTimeBackup;
```

Comment:
What is the purpose of the backup value?

Reply:
Some of the mappings can use either effectiveTime.center or effectiveTime.low

Reply:
yes, I understand it now, thanks